File: plugin/graph/src/main/groovy/net/twisterrob/gradle/graph/javafx/interop/TaskResultSerializer.java
package net.twisterrob.gradle.graph.javafx.interop;

import java.lang.reflect.Type;
import java.util.EnumMap;

import com.google.gson.*;

import net.twisterrob.gradle.graph.TaskResult;

public class TaskResultSerializer implements JsonSerializer<TaskResult> {
	@Override public JsonElement serialize(TaskResult taskResult, Type type, JsonSerializationContext context) {
		String stateString = getState(taskResult);
		return stateString != null? new JsonPrimitive(stateString) : null;
	}

	private static final EnumMap<TaskResult, String> MAPPING = new EnumMap<>(TaskResult.class);

	static {
		MAPPING.put(TaskResult.executing, "executing");
		MAPPING.put(TaskResult.nowork, "nowork");
		MAPPING.put(TaskResult.skipped, "skipped");
		MAPPING.put(TaskResult.uptodate, "uptodate");
		MAPPING.put(TaskResult.failure, "failure");
		MAPPING.put(TaskResult.completed, "success");
		assert MAPPING.keySet().size() == TaskResult.values().length;
	}

	public static String getState(TaskResult result) {
		return MAPPING.get(result);
	}
}

File: graph/src/main/kotlin/net/twisterrob/gradle/graph/vis/d3/interop/TaskResultSerializer.kt
package net.twisterrob.gradle.graph.vis.d3.interop

import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import net.twisterrob.gradle.graph.tasks.TaskResult
import java.lang.reflect.Type
import java.util.EnumMap

class TaskResultSerializer : JsonSerializer<TaskResult> {

	override fun serialize(taskResult: TaskResult, type: Type, context: JsonSerializationContext): JsonElement =
		JsonPrimitive(getState(taskResult))

	companion object {

		private val MAPPING: Map<TaskResult, String> =
			EnumMap<TaskResult, String>(TaskResult::class.java).apply {
				this[TaskResult.Executing] = "executing"
				this[TaskResult.NoWork] = "nowork"
				this[TaskResult.Skipped] = "skipped"
				this[TaskResult.UpToDate] = "uptodate"
				this[TaskResult.NoSource] = "nosource"
				this[TaskResult.FromCache] = "fromcache"
				this[TaskResult.Failure] = "failure"
				this[TaskResult.Completed] = "success"
				check(this.keys.size == TaskResult.values().size)
			}

		fun getState(result: TaskResult): String =
			MAPPING.getValue(result)
	}
}
